src/main/java/org/folio/rest/impl/InstanceStorageApi.java

```diff
@@ -387,6 +389,31 @@ public void putInstanceStorageInstancesSourceRecordModsByInstanceId(
         .respond500WithTextPlain("Not implemented yet.")));
   }
 
+  @Validate
+  @Override
+  public void postInstanceStorageInstancesRetrieve(RetrieveDto entity,
+                                                   RoutingContext routingContext,
+                                                   Map<String, String> okapiHeaders,
+                                                   Handler<AsyncResult<Response>> asyncResultHandler,
+                                                   Context vertxContext) {
+    if (PgUtil.checkOptimizedCQL(entity.getQuery(), TITLE) != null) {
+      try {
+        PreparedCql preparedCql = handleCql(entity.getQuery(), entity.getLimit(), entity.getOffset());
+        PgUtil.getWithOptimizedSql(preparedCql.getTableName(), Instance.class, Instances.class,
+          TITLE, entity.getQuery(), entity.getOffset(), entity.getLimit(),
+          okapiHeaders, vertxContext, GetInstanceStorageInstancesResponse.class, asyncResultHandler);
+      } catch (Exception e) {
+        log.error(e.getMessage(), e);
+        asyncResultHandler.handle(io.vertx.core.Future.succeededFuture(
+          GetInstanceStorageInstancesResponse.respond500WithTextPlain(e.getMessage())));
+      }
+      return;
+    }
+
+    PgUtil.streamGet(INSTANCE_TABLE, Instance.class, entity.getQuery(), entity.getOffset(), entity.getLimit(), null,
+      "instances", routingContext, okapiHeaders, vertxContext);
```

Comment:
It's pretty same as in `getInstanceStorageInstances` could you extract common code?